DynamicLinking.md

```diff
@@ -1,49 +1,87 @@
-on many aspects of the design that do impact the MVP, such as the way linear
-memory is managed, how module imports and exports are specified, and how globals
-and function pointers work.  Therefore we want to have some viable ideas to
-ensure we don't standardize a design that unnecessarily complicates the design
-or implementation of dynamic linking.
-
-  [hosted libraries]: https://developers.google.com/speed/libraries/
-  [service workers]: https://www.w3.org/TR/service-workers/
+WebAssembly enables load-time and run-time (`dlopen`) dynamic linking in the
+MVP by having multiple [instantiated modules](Modules.md)
+share functions, [linear memories](AstSemantics.md#linear-memory),
+[tables](AstSemantics.md#table) and [constants](AstSemantics.md#constants)
+using module [imports](Modules.md#imports) and [exports](Modules.md#exports). In
+particular, since all (non-local) state that a module can access can be imported
+and exported and thus shared between separate modules' instances, toolchains
+have the building blocks to implement dynamic loaders.
+
+Since the manner in which modules are loaded and instantiated is defined by the
+host environment (e.g., the [JavaScript API](JS.md)), dynamic linking requires
+use of host-specific functionality to link two modules. At a minimum, the host
+environment must provide a way to dynamically instantiate modules while
+connecting exports to imports.
+
+The simplest load-time dynamic linking scheme between modules A and B can be 
+achieved by having module A export functions, tables and memories that are
+imported by B. A C++ toolchain can expose this functionality by using the
+same function attributes currently used to export/import symbols from
+native DSOs/DLLs:
+```
+#ifdef _WIN32
+#  define EXPORT __declspec(dllexport)
+#  define IMPORT __declspec(dllimport)
+#else
+#  define EXPORT __attribute__ ((visibility ("default")))
+#  define IMPORT __attribute__ ((visibility ("default")))
+#endif
+
+typedef void (**PF)();
+
+IMPORT PF imp();
+EXPORT void exp() { (*imp())(); }
+```
+This code would, at a minimum, generate a WebAssembly module with imports for:
+* the function `imp`
+* the heap used to perfom the load, when dereferencing the return value of `imp`
+* the table used to perform the pointer-to-function call
+
+and exports for:
+* the function `exp`
+
+A more realistic module using libc would have more imports including:
+* an immutable `i32` global import for the offset in linear memory to place
+  global [data segments](Modules.md#data-section) and later use as a constant
+  base address when loading and storing from globals
+* an immutable `i32` global import for the offset into the indirect function
+  table at which to place the modules' indirectly called functions and later
+  compute their indices for address-of
+
+One extra detail is what to use as the [module name](Modules.md#imports) for
+imports (since WebAssembly has a two-level namespace). One option is to have a
```

Comment:
Is there a reference for the two-level namespace stuff? I can't seem to find it, maybe just file a bug to add it later so the PR doesn't get bogged down?